### ISSUE TITLE
What happens on the device timeline after `GPUDevice.destroy` is called?

### ISSUE THREAD

**Issue by jimblandy** [gpuweb/gpuweb#4294]
The `GPUDevice.destroy` method doesn't mark the device invalid until we **lose the device**, which happens after "all currently-enqueued operations on any queue on this device are completed". Are other operations allowed to occur on the device timeline while the `destroy` operation waits for this?

The `destroy` method's specification is split into two device-timeline sections, with the above phrase seeming to say when the second section begins execution. Since we only mark the device invalid in **lose the device**, this suggests to me that other work can happen on the device timeline between those two sections of `destroy`.

This means that an operation that doesn't touch the queue timeline, like buffer creation, could fully complete both its content and device timeline work *after* a `destroy` call returns to the content timeline. This is surprising.

Am I reading the spec correctly? Is this intentional?

**Comment by Kangz** [comment]
I believe this is unintentional, the device destruction should immediately prevent all future uses of the device, including creating objects.

**Comment by kainino0x** [comment]
Intent is that after device destroy, any attempt to do anything does nothing and returns no errors (except for mapping, where we reject the promise). #1629

meeting: Move to editorial to improve the language (at least add non-normative notes even if everything else is technically correct already - due to everything being unobservable after device loss/destruction).

**Comment by kainino0x** [comment]
I started looking into fixing this, and it's harder than I realized. The reason that `device.destroy()` waits for queue operations to finish was #2599, for which we wanted to flush the queue and finish any pending maps before actually losing the device.

The best/only idea I can think of is to introduce a "destroy pending" state so that no new operations can be injected after destruction starts.

Unfortunately I don't think we can rely entirely on non-observability to say implementations are allowed to do that internally. It would work for `createBuffer` on its own (you can't tell createBuffer has finished unless you map it or something), but it wouldn't work for `mapAsync` which could in principle start and finish before `all currently-enqueued operations on any queue on this device are completed`.

Fortunately only the 6 Promise-returning functions can actually send signals back to the content timeline, so I am pretty sure only those 6 would need to actually check the "destroy pending" state, and others could rely on non-observability.

**Comment by teoxoy** [comment]
Couldn't device destroy set the devices's valid property to false as the first step on the device timeline? I think that would work since the relevant API calls already check for the device to be valid.

**Comment by kainino0x** [comment]
That would cause the mapAsync()'s second device-timeline block to see the device is lost, and reject the promise, which breaks #2599.

I should add something that explicitly says that steps that were waiting on queue completion events will run before the device.detsroy()'s second device-timeline block.

**Comment by teoxoy** [comment]
`Device.destroy()` unmaps all mapped buffers as a first step though. The 2nd device timeline block of `mapAsync` looks unreachable to me if you call `Device.destroy()`.

**Comment by teoxoy** [comment]
> Do some work, issue a mapAsync for readback, and destroy before waiting for the map to complete.

This sound like the underlying Buffer would still have to be alive and mapped even after all steps of `Device.destroy()` completed. This seems quite hard to implement (unless you copy the whole mapped region), Vulkan requires all resources created by the device to be destroyed before the device is.

**Comment by teoxoy** [comment]
@kainino0x thoughts on this? ^

**Comment by kainino0x** [comment]
> `Device.destroy()` unmaps all mapped buffers as a first step though. The 2nd device timeline block of `mapAsync` is reachable but the promise was already rejected by `unmap`.

You're totally right, I had not connected the dots between these things. So the spec doesn't even allow the map-and-destroy thing to work right now, because canceling the maps makes it unobservable. Convenient! (Even if it were possible, I agree it very possibly would be too much to ask of implementations to make that work.)

As for render-to-canvas-and-destroy, unfortunately we can't hide behind canceling async work because canvases are visible to the user and can also be read back synchronously. We can could allow it to be racy (the final steady-state result depends on when the destroy() call makes it to the device) but if we want to avoid that we have to either:

- Allow it to be racy: the final canvas contents depend on when the destroy() call makes it to the device, potentially displaying a partial frame because destruction happened in between queue operations. Seems not good.
- Wait for queue completion like we do now - the canvas is left with the final rendered result. This could be difficult to implement depending on how the browser allocates the canvas buffer (inside or outside the webgpu device).
- Clear the canvas when destroying the device. This would still be racy for ~1 frame, but the steady-state result would be a blank canvas.

**Comment by kainino0x** [comment]
Oh, we should think about device loss as well. I think it's pretty uncommon to have a natural device loss that's scoped to a single WebGPU device and doesn't take down the whole GPU process, but if that _were_ to happen, then it could also result in displaying a partially-rendered canvas frame like (1). Of course (2) is not possible in that case. (3) is doable, we could cover both cases by having the canvas contents return blank if the canvas's device is lost.

I should note (3) not only prevents destroying immediately after submitting rendering work, but also destroying _later_ even after the canvas has been successfully rendered and displayed. Destroying will still blank the canvas. This is probably OK....

**Comment by kainino0x** [comment]
> This sound like the underlying Buffer would still have to be alive and mapped even after all steps of `Device.destroy()` completed. This seems quite hard to implement (unless you copy the whole mapped region), Vulkan requires all resources created by the device to be destroyed before the device is.

Actually, the natural device loss scenario hits this too, for buffers that are already mapped. In that case the device is lost before we have any opportunity to revoke JS's map access. This is fine (for chromium) today\* because the content process never gets a pointer to the actual mapped memory, just interprocess shared memory. And I think it would also be fine for explicitly destroying devices too, if not for the fact that we already unmap/abort mappings. (Multithreading also introduces another scenario, as destroy() from one thread can't terminate existing mappings on other threads.)

\* IIRC in the "triple mapping" scheme (where we mmap the actual mapped memory into the content process), that memory gets allocated at the OS level (and imported to the driver), not allocated at the driver level, so it would stay alive after the device is lost.

**Comment by kainino0x** [comment]
I tentatively added this to the meeting agenda in case we need to discuss it.

I would like to keep the original decision from #2599 as it applies to canvas rendering. (It no longer applies to mapping since we cancel same-thread maps on destroy, but it will again with multithreading, so I think the spec change I was working on is still valid.)

There is a conformance test (reftest) that canvases retain their contents after the device is destroyed:
<https://gpuweb.github.io/cts/out/webgpu/web_platform/reftests/canvas_display_after_device_lost.https.html> (and some more here that I never got around to landing: https://github.com/gpuweb/cts/pull/3715)
It doesn't wait for drawing to complete before destroying, though the queue work is very fast so it doesn't try very hard to test that we're flushing the queue before destroy. It could be enhanced with cases that have queue work that takes longer, I think.

**Comment by kainino0x** [comment]
Filed #4859

**Comment by Kangz** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/GPU-Web-2024-09-04#m0-what-happens-on-the-device-timeline-after-gpudevicedestroy-is-called-4294">GPU Web WG 2024-09-04 Atlantic-time</a></summary>

* JB: nice to have a summary of where you landed.
* KN: my question was: can you render to a canvas and then destroy the device? 2 questions. 1) If you destroy the device before it finishes, will it display? 2) If you destroy after rendering, will the canvas stay on the screen, or will it be cleared? Discussed in the past. If you destroy, it'll finish what you rendered, and canvas contents will be preserved. Works for us in Chromium because we allocate the canvas's texture outside the canvas, but can imagine has impl complexity for others. We have a reftest already for this. Would like to enhance the test - do a big draw, destroy the device.
* JB: that's what you meant by keeping #2599.
* KN: yes. 1) mapping and destroying; 2) rendering to canvas and destroying. We decided destroying cancels the map. Only relevant thing is the canvas.
* JB: our general feeling: wasn't our intention to revisit #2599. Only thing we're concerned about, preventing absurd things like creating buffers on the destroyed device. Destroy pending state -> ack that work's still ongoing, but prevent device from being used by further operations. How to spec, leave up to the editors.
* KN: did land the destroyStarted change. Spec as-is is what I described in keeping the original decision.
* JB: OK, then maybe no further work here.
* KN: agree
* MW: thumbs up
* KG: want as few states as possible. Nice to have device.destroy transition immediately on the client. Other thing, deliberately drop the canvas and then destroy the device. Device destruction happens before presentation to HTML page. Do we take that partially- or completely-rendered back buffer as a fully rendered frame, if device went invalid during that rAF?
* JB: that's the #2599 issue?
* KG: don't think so.
* KN: I get the concern. We have a case like this in our impl. Our backbuffer's a WebGPU texture, then later we copy into the canvas, rather than rendering directly into the canvas. If the copy happens after the destroy, won't work - src texture isn't there any more. Spec has non-normative notes about this.
* KG: don't care that much, just worrying we're adding complexity that we'll regret.
* JB: worth going to some effort to offer predictable behavior. Want API to behave similarly regardless of how quickly the GPU finishes its work.
* KG: that's understood.
* KN: there's a "wait for the queue to finish" part in our spec. Alternative is, assuming we want consistent behavior: once you destroy the device, the canvas goes blank. Then doesn't matter when it happens.
* CW: isn't that problematic if the device gets lost?
* KG: in the case of a lost canvas you try your best to preserve the content. But if you try to preserve during device destruction, may get a different result than that.
* GT: if you draw something and then wait 10 seconds and then destroy, would expect to see the previous draw.
* BJ: think it depends on whether you've called getCurrentTexture or not. Fetched a new texture? We'll clear the canvas and process things.
* KN: think spec says you'll get the last thing you rendered.
* CW: submitted.
* KN/BJ: yes.
* KG: queue submitted? OK. Problem is: queue submission doesn't put things on the screen. Presentation to HTML document does. That's a different step.
* CW: rewinding: textures returned by getCurrentTexture on most browsers are "externally allocated" from the WebGPU implementation. Lent to the WebGPU impl for the duration of getCurrentTexture until end of rAF / next getCurrentTexture. If you bind that to a texture, queue.submit, device.destroy - pipeline of commands to GPU process will do the IOSurface render before destruction. Then end of rAF will display it.
* KN: does presentation involve the device? If so, can't do it. Think it's nice behavior - smooths over displaying incomplete stuff because of device loss. Have to finish the work you just submitted. I'm interested in this idea - that getCurrentTexture removes the canvas back buffer, and end of rAF, if the device is still alive, takes the texture and puts it on the canvas.
* KN: not sure if we need to clear the canvas or not. Presentation would fail. What happens? 1) retain previous frame 2) get a blank one, because you called getCurrentTexture. Both are sound, both are nicer than what I've been trying to propose.
* CW: talked a lot. Kai you're volunteering to write a new issue about this?
* KR: personal opinion, like the canvas going blank. It's simpler. Device loss is pretty much impossible to preserve the canvas's contents.
* More discussion.
* BJ: for later: think it's important for developers to have a guaranteed way to render one frame. As long as there's a consistent way to do this (and we do, through ImageBitmap), just need to document that well.

</details>
